runtime/compiler/runtime/J9Profiler.cpp

```diff
@@ -1763,6 +1764,7 @@ TR_BlockFrequencyInfo::getFrequencyInfo(
          callStack.push_back(std::make_pair(comp->fe()->getInlinedCallSiteMethod(callSite), bciToCheck));
          bciToCheck = callSite->_byteCodeInfo;
          }
+      callStack.push_back(std::make_pair(comp->getCurrentMethod()->getPersistentIdentifier(), bciToCheck));
```

Comment:
Hmm, looking at this again, I think that if we push here, it will mess with the loop below, e.g. empty doesn't mean the same thing, `bciToCheck` would need to start coming from the second-last element, etc. But rather than make all those adjustments, I think we can maintain the stack exactly as before, but push the extra element temporarily just for `computeEffectiveCallerIndex()`. I think it would work to just skip this push, and then later:
```diff
 while (!callStack.empty())
    {
-   bciToCheck = callStack.back().second;
+   auto extraCaller = callStack.back();
+   bciToCheck = extraCaller.second;
    callStack.pop_back();
```
and:
```diff
 int32_t effectiveCallerIndex = -1;
 TR_BlockFrequencyInfo *bfi = info->getBlockFrequencyInfo();
- if (callStack.empty() || info->getCallSiteInfo()->computeEffectiveCallerIndex(comp, callStack, effectiveCallerIndex))
+ bool computeFrequency = callStack.empty();
+ if (!computeFrequency)
+   {
+   callStack.push_back(extraCaller);
+   computeFrequency = info->getCallSiteInfo()->computeEffectiveCallerIndex(comp, callStack, effectiveCallerIndex);
+   callStack.pop_back();
+   }
+
+ if (computeFrequency)
    {
    TR_ByteCodeInfo callee(bci);
    callee.setCallerIndex(effectiveCallerIndex);
```
That way the only difference is that within `computeEffectiveCallerIndex()` we have the one extra element at the end, and that's exactly what's needed, since its loop will stop one earlier, and it will only look at the last element in order to compare the bytecode index

Reply:
Thanks for the patch, updated in https://github.com/eclipse-openj9/openj9/pull/16205/commits/c154eb386acafddf5a31848aa278c7f41aa69d05

Reply:
Thanks! I still think this particular push (at the top of the thread) should go though. All it does is add an extra (first) iteration to the loop below, and that extra iteration will check whether the entire stack matches, but we already know it doesn't match, which is why we're looking for a grafting point

Reply:
Yes, sorry, missed to remove this 😅 , fixed in https://github.com/eclipse-openj9/openj9/compare/c154eb386acafddf5a31848aa278c7f41aa69d05..31808dc7926174e799184bb01decbf6caad0aacd